block-layer/codecs/bitcoin.md

```diff
@@ -0,0 +1,220 @@
+
+Navigating using IPLD linking from a Bitcoin block graph header to transactions requires traversal via the "transaction Merkle root" through binary Merkle tree to the desired transaction. Navigating to individual transactions without these internal binary Merkle tree nodes is not possible using IPLD's linking system, which is why they must be present in explicit form when viewed with IPLD tooling.
+
+## Bitcoin Block Graph Binary Format
+
+The Bitcoin block graph binary format, available as hexadecimal via the Bitcoin Core CLI using the `getblock <address> 0` command, includes the following elements:
+
+1. Header - the first 80 bytes
+2. A list of one or more Transactions - for each Transaction:
+  2.1 A list of one or more Transaction Ins - detailing historical transactions (and a specific Transaction Out) where coins are being transferred from, can be used to derive the wallet to debit
+  2.2 A list of one or more Transaction Outs - detailing wallets where coins are being transferred (credited) to
+3. Transaction witness data - directly after some transactions, since [BIP141 (SegWit)](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki)
+
+### Header
+
+The first 80 bytes of the raw format contains the Header. The double-SHA2-256 digest of these 80 bytes forms what is commonly known as the "block address" and begins with a number of zeros, which increases over time as mining difficulty increases.
+
+Decorated with links, the basic structure of the Header is as follows:
+
+```ipldsch
+type Header struct {
+  version Int
+  previousblockhash Bytes
+  parent &Header # previousblockhash converted to CID with `bitcoin-block` codec
+  merkleroot Bytes
+  tx &TransactionMerkle # merkleroot converted to CID with `bitcoin-tx` codec
+  time Int
+  bits Int
+  nonce Int
+}
+```
+
+Decoding the 80 bytes into this structure requires reading the following elements in order:
+
+* `version`: a signed 32-bit integer
+* `previousblockhash`: an unsigned 256-bit integer / a 32-byte binary string
+* `merkleroot`: an unsigned 256-bit integer / a 32-byte binary string
+* `time`: an unsigned 32-bit integer
+* `bits`: an unsigned 32-bit integer
+* `nonce`: an unsigned 32-bit integer
+
+Once decoded, the Header may be represented in its full porcelain form, as provided by the Bitcoin Core RPC in JSON:
+
+```ipldsch
+type Block struct {
+  hash String                       # uint256 BE in hex form, the double-SHA2-256 of the block header
+  version Int
+  versionHex String                 # same as version, but base16
+  previousblockhash optional String # uint256 BE in hex form, will not be present for the genesis block
+  merkleroot String                 # uint256 BE in hex form
+  time Int
+  bits Int
+  difficulty Float                  # derived from `bits`
```

Comment:
OK, I need to work on the language around this more, but the lead-in to this particular snippet of Schema says "the Header **may be represented**". There's no binary involved in `difficulty`, it's purely something that's _only_ present in the Data Model, just like `versionHex`. It's also not critical that an implementation of an IPLD Bitcoin codec include this since a consumer could derive it for themselves. I've done it so I can generate almost exactly matching JSON to compare against the Bitcoin Core RPC's output to validate my data.

There's more of this kind of awkwardness coming up, particularly when it comes to script decoding! When it comes to codecs like Bitcoin, what do we present at the Data Model layer? What does it mean to "decode" Bitcoin data? Do we _only_ present exactly what we find in the bytes? Does that mean we leave out the implicit Links (I'm inserting a `tx` CID to link to transactions and leaving `merkleroot` as a byte array, so it's essentially duplicated but not navigable in the raw form)? There's a bunch of different points along the "decode" spectrum that can be chosen and I think I'll have to write that up below and leave it to implementers to decide how far along they go.

I went the whole way with Bitcoin, I can even fully decode scripts into their string form and present them as the Bitcoin Core RPC does, so my decode/encode validation goes both ways and compares against the authoritative and only needs to discount the few chain-contextual pieces of data (`chainwork` etc.). Even as I'm working on Zcash (again), I think there's parts of it where I'm probably going to stop short and shortcut my validation because there's definitely diminishing value as you get into more detail.